wrangler-storage/src/main/java/io/cdap/wrangler/store/upgrade/UpgradeStore.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package io.cdap.wrangler.store.upgrade;
+
+import io.cdap.cdap.api.NamespaceSummary;
+import io.cdap.cdap.spi.data.StructuredRow;
+import io.cdap.cdap.spi.data.StructuredTable;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.table.StructuredTableId;
+import io.cdap.cdap.spi.data.table.StructuredTableSpecification;
+import io.cdap.cdap.spi.data.table.field.Field;
+import io.cdap.cdap.spi.data.table.field.Fields;
+import io.cdap.cdap.spi.data.table.field.Range;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Upgrade store to store upgrade information for each namespace,
+ * Use system namespace to track the overall upgrade status
+ */
+public class UpgradeStore {
+  private static final NamespaceSummary SYSTEM_NS = new NamespaceSummary("system", "", 0L);
+  private static final StructuredTableId TABLE_ID = new StructuredTableId("wrangler_upgrade");
+
+  private static final String NAMESPACE_COL = "namespace";
+  private static final String GENERATION_COL = "generation";
+  private static final String CONNECTION_UPGRADE_COMPLETE_FIELD = "connection_upgrade_complete";
+  private static final String WORKSPACE_UPGRADE_COMPLETE_FIELD = "workspace_upgrade_complete";
+  private static final String UPGRADE_TIMESTAMP = "upgrade_timestamp";
+
+  public static final StructuredTableSpecification UPGRADE_TABLE_SPEC =
+    new StructuredTableSpecification.Builder()
+      .withId(TABLE_ID)
+      .withFields(Fields.stringType(NAMESPACE_COL),
+                  Fields.longType(GENERATION_COL),
+                  Fields.stringType(CONNECTION_UPGRADE_COMPLETE_FIELD),
```

Comment:
We should treat the UpgradeStore as something that can be re-used for future upgrades without needing to update the table schema or creating yet another table.

I think it would be better to have a column for the entity type (and add it to the primary key), with 'workspace' and 'connection' as possible values today. Have another column that contains the current state of that entity, which could be a serialized UpgradeState object which includes a version or something like that.

This way, it could be re-used for other purposes, such as upgrading drafts or other application specific data.